sys: lists: Add list functions to compute list length

J-Montgomery opened the pull request:
Many parts of zephyr regularly need to compute the length of various lists, as indicated in #42775
This commit adds functions to slist, sflist, and dlist to return the length of provided lists, which will allow future commits to refactor code that implements this manually. The commit also adds tests for these APIs to the corresponding test cases. It is noted that the test functions calculate equivalent values manually using several different internal list functions. This has been left unmodified to ensure that each of the various ways to traverse the list compute the same value as the new list_len() functions.

The new functions all take a reference to any node in the list and compute the length of the entire list. If the list is empty, they return 0.

MaureenHelm commented:
@andyross please take a look

cfriedt reviewed (changes requested):
@J-Montgomery - This looks reasonable to me. Only 1 change request. Can you please create a separate second commit that includes the changes for the testsuite? It's just a Zephyr convention to have separate test and code commits.

andyross reviewed (dismissed):
Looks fine to me.  Implemented cleanly in the obvious way, and even honors the somewhat weird GENLIST nonsense we do to enable the flag list.

Agree that the test code should be split to a separate patch.

Also one bit of pedantry that would be nice to see would be for the docs to explicitly call out the runtime.  These aren't fancy data structures that contain a list count, the *_len() predicates compute the length by walking the full list, and are thus O(N) in the length of the list.

Dropping a +1 on it in advance of those changes, knowing that @cfriedt's -1 will prevent early merge.

J-Montgomery committed:
sys: lists: Add list functions to compute list length

This commit adds functions to slist, sflist, and dlist to return the
length of provided lists, which will allow future commits to refactor
code that implements this manually.

The new functions all take a reference to any node in the list and compute
the length of the entire list. If the list is empty, they return 0.

Signed-off-by: J M <montytyper@msn.com>

J-Montgomery committed:
sys: lists: Add tests for newly-added list len functions

This commit adds tests for the newly added list len APIs to the
corresponding test cases. It is noted that the test functions calculate
equivalent values manually using several different internal list
functions. This has been left unmodified to ensure that the manual ways
using each of the various for_each functions results in the same value
as the new list_len() functions.

Signed-off-by: J M <montytyper@msn.com>

J-Montgomery commented:
Apologies for taking so long to get around to this, been dealing with a wedding. Requested changes are made, including the docs update.

cfriedt commented:
Just going to hold off on approval until after v3.4.0 is tagged. Thanks for your patience.

cfriedt commented:
@andyross - refresh +1 if you don't mind

OneTone commented on include/zephyr/sys/list_gen.h:
```diff
@@ -234,4 +234,15 @@
+	static inline size_t sys_##__lname##_len(sys_##__lname##_t * list)                         \
+	{                                                                                          \
+		size_t len = 0;                                                                    \
+		static sys_##__nname##_t * node;                                                   \
```
I hope it's okay to comment even though the PR has already been merged for a long time.  

I have a question: why is the iterator variable `node` defined as `static` linkage?

cfriedt replied:
Good question. @OneTone - if you make a PR to fix it, I would approve

OneTone replied:
Thank you @cfriedt !  
I've just created a PR to address the issue: #76973 .  
It's my first contribution to this project, so please feel free to point out any problems.  
I'd really appreciate your feedback!
